src/generic/FreeAssAlgebraGroebner.jl

```diff
@@ -0,0 +1,626 @@
+###############################################################################
+#
+#   FreeAssAlgebraGroebner.jl : free associative algebra R<x1,...,xn> Groebner basis
+#
+###############################################################################
+
+export groebner_basis
+
+export interreduce!
+
+export normal_form
+
+export normal_form_weak
+
+import DataStructures: PriorityQueue, enqueue!, dequeue!
+
+const groebner_debug_level = 0
+const Monomial = Vector{Int}
+
+abstract type Obstruction{T} end
+"""
+represents the overlap of the leading term of two polynomials
+"""
+struct ObstructionTriple{T} <: Obstruction{T}
+    first_poly::FreeAssAlgElem{T}
+    second_poly::FreeAssAlgElem{T}
+    pre_and_suffixes::NTuple{4,Monomial}
```

Comment:
Why is this a 4-tuple instead of four separate fields? There may be a good reason but w/o knowing what the entries even are, this is hard to understand. 

Another option would be to turn this into its own struct, or use a `NamedTuple`